store/store.go

```diff
@@ -0,0 +1,519 @@
+		basepath:     basePath,
+		cache:        cache.NewDoubleCache(recentBlocksCache, blockstoreCache),
+		stripLock:    newStripLock(1024),
+		emptyHeights: emptyHeights,
+	}
+	return store, nil
+}
+
+func (s *Store) Close() error {
+	return s.storeEmptyHeights()
+}
+
+func (s *Store) Put(
+	ctx context.Context,
+	datahash share.DataHash,
+	height uint64,
+	square *rsmt2d.ExtendedDataSquare,
+) (eds.AccessorStreamer, error) {
+	tNow := time.Now()
+	lock := s.stripLock.byDatahashAndHeight(datahash, height)
+	lock.lock()
+	defer lock.unlock()
+
+	if datahash.IsEmptyRoot() {
+		s.addEmptyHeight(height)
+		return emptyFile, nil
+	}
+
+	// short circuit if file exists
+	if has, _ := s.hasByHeight(height); has {
+		s.metrics.observePutExist(ctx)
+		return s.getByHeight(height)
+	}
+
+	filePath := s.basepath + blocksPath + datahash.String()
+	f, err := s.createFile(filePath, datahash, square)
+	if err != nil {
+		s.metrics.observePut(ctx, time.Since(tNow), square.Width(), true)
+		return nil, fmt.Errorf("creating file: %w", err)
+	}
+
+	// create hard link with height as name
+	err = s.createHeightLink(datahash, height)
+	if err != nil {
+		s.metrics.observePut(ctx, time.Since(tNow), square.Width(), false)
+		return nil, fmt.Errorf("linking height: %w", err)
+	}
+	s.metrics.observePut(ctx, time.Since(tNow), square.Width(), false)
+
+	// put file in recent cache
+	f, err = s.cache.First().GetOrLoad(ctx, height, fileLoader(f))
+	if err != nil {
+		log.Warnf("failed to put file in recent cache: %s", err)
+	}
+	return f, nil
+}
+
+func (s *Store) createFile(
+	filePath string,
+	datahash share.DataHash,
+	square *rsmt2d.ExtendedDataSquare,
+) (eds.AccessorStreamer, error) {
+	// check if file with the same hash already exists
+	f, err := s.getByHash(datahash)
+	if err == nil {
+		return f, nil
+	}
+
+	if !errors.Is(err, ErrNotFound) {
+		return nil, fmt.Errorf("getting by hash: %w", err)
+	}
+
+	// create Q1Q4 file
+	f, err = file.CreateQ1Q4File(filePath, datahash, square)
+	if err != nil {
+		return nil, fmt.Errorf("creating ODS file: %w", err)
+	}
+	return f, nil
+}
+
+func (s *Store) GetByHash(ctx context.Context, datahash share.DataHash) (eds.AccessorStreamer, error) {
+	if datahash.IsEmptyRoot() {
+		return emptyFile, nil
+	}
+	lock := s.stripLock.byDatahash(datahash)
+	lock.RLock()
+	defer lock.RUnlock()
+
+	tNow := time.Now()
+	f, err := s.getByHash(datahash)
+	s.metrics.observeGet(ctx, time.Since(tNow), err != nil)
+	return f, err
+}
+
+func (s *Store) getByHash(datahash share.DataHash) (eds.AccessorStreamer, error) {
+	if datahash.IsEmptyRoot() {
+		return emptyFile, nil
+	}
+
+	path := s.basepath + blocksPath + datahash.String()
+	f, err := file.OpenQ1Q4File(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrNotFound
+		}
+		return nil, fmt.Errorf("opening ODS file: %w", err)
+	}
+	return wrappedFile(f), nil
+}
+
+func (s *Store) LinkHashToHeight(_ context.Context, datahash share.DataHash, height uint64) error {
```

Comment:
Why do we need this public?

Reply:
It will be used to short circuit add only link, if hash exists. Will be usefull in Availability, when we know datahash and heigh of DAH, so we can avoid doing network request and just add height link.